File: Manuel/개인 활동/Class 3/DFS와 BFS/1260.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#define endl "\n"

using namespace std;

void dfs(const vector<vector<int>>& graph, int v, vector<bool>& visited, vector<int>& result)
{
    visited[v] = true;
    result.push_back(v);
    for(int neighbour : graph[v])
    {
        if(!visited[neighbour])
        {
            dfs(graph, neighbour, visited, result);
        }
    }
}

void bfs(const vector<vector<int>>& graph, int v, vector<bool>& visited, vector<int>& result)
{
    queue<int> q;
    q.push(v);
    visited[v] = true;
    result.push_back(v);

    while(!q.empty())
    {
        int current = q.front();
        q.pop();
        
        for(int neighbour : graph[current])
        {
            if(!visited[neighbour])
            {
                visited[neighbour] = true;
                q.push(neighbour);
                result.push_back(neighbour);
            }
        }
    }
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int N, M, V;
    cin >> N >> M >> V;
    vector<vector<int>> graph(N + 1);

    for(int i = 0; i < M; i++)
    {
        int n1, n2;
        cin >> n1 >> n2;
        graph[n1].push_back(n2);
        graph[n2].push_back(n1);
    }
    for(int i = 1; i <= N; i++)
    {
        sort(graph[i].begin(), graph[i].end());
    }

    vector<bool> dfs_visited(N + 1, false);
    vector<int> dfs_result;
    dfs(graph, V, dfs_visited, dfs_result);
    for(size_t i = 0; i < dfs_result.size(); i++)
    {
        cout << dfs_result[i] << (i == dfs_result.size() - 1 ? "" : " ");
    }
    cout << endl;

    vector<bool> bfs_visited(N + 1, false);
    vector<int> bfs_result;
    bfs(graph, V, bfs_visited, bfs_result);
    for(size_t i = 0; i < bfs_result.size(); i++)
    {
        cout << bfs_result[i] << (i == bfs_result.size() - 1 ? "" : " ");
    }
    cout << endl;

    return 0;
}

File: Manuel/개인 활동/Class 3/DFS와 BFS/1260.py
import sys
from collections import deque

def dfs(graph, v, visited, result):
    visited[v] = True
    result.append(v)
    for neighbour in graph[v]:
        if not visited[neighbour]:
            dfs(graph, neighbour, visited, result)

def bfs(graph, v, visited, result):
    q = deque([v])
    visited[v] = True
    result.append(v)
    while q:
        current = q.popleft()
        for neighbour in graph[current]:
            if not visited[neighbour]:
                visited[neighbour] = True
                q.append(neighbour)
                result.append(neighbour)

def solve():
    N, M, V = map(int, sys.stdin.readline().split())
    graph = [[] for _ in range(N + 1)]

    for _ in range(M):
        n1, n2 = map(int, sys.stdin.readline().split())
        graph[n1].append(n2)
        graph[n2].append(n1)
    
    for i in range(1, N + 1):
        graph[i].sort()

    dfs_visited = [False] * (N + 1)
    dfs_result = []
    dfs(graph, V, dfs_visited, dfs_result)
    print(*dfs_result)

    bfs_visited = [False] * (N + 1)
    bfs_result = []
    bfs(graph, V, bfs_visited, bfs_result)
    print(*bfs_result)

solve()